os/drivers/lcd/mipi_lcd.c

```diff
@@ -48,34 +53,159 @@ struct mipi_lcd_dev_s {
-	} else if (payload_len == 1) {
-		msg.type = MIPI_DSI_DCS_SHORT_WRITE_1_PARAM;
-	} else {
-		msg.type = MIPI_DSI_DCS_LONG_WRITE;
-	}
-	msg.tx_buf = cmd_addr;
-	msg.tx_len = payload_len;
+	msg.channel = 0xB0;
+	msg.type = MIPI_DSI_DCS_LONG_WRITE;
+	msg.tx_buf = cmd_mode;
+	msg.tx_len = 2;
 	msg.flags = 0;
+	transfer_status = mipi_dsi_transfer(priv->dsi_dev, &msg);
+	if (transfer_status != OK) {
+		lcddbg("Command %x not sent \n", msg.channel);
+	}
+	priv->mode = COMMAND;
+	return transfer_status;
+}
+
+static int switch_to_video(struct mipi_lcd_dev_s *priv)
+{
+	int transfer_status = OK;
 	priv->config->lcd_mode_switch(false);
+	uint8_t cmd_mode[2] = {0x00, 0x10};
+	struct mipi_dsi_msg msg;
+	msg.channel = 0xB0;
+	msg.type = MIPI_DSI_DCS_LONG_WRITE;
+	msg.tx_buf = cmd_mode;
+	msg.tx_len = 2;
+	msg.flags = 0;
 	transfer_status = mipi_dsi_transfer(priv->dsi_dev, &msg);
+	if (transfer_status != OK) {
+		lcddbg("Command %x not sent \n", msg.channel);
+	}
 	priv->config->lcd_mode_switch(true);
+	priv->mode = VIDEO;
+	return transfer_status;
+}
+
+static int send_cmd(struct mipi_lcd_dev_s *priv, bool display_on)
+{
+	priv->config->lcd_mode_switch(false);
+	int transfer_status = OK;
+	struct mipi_dsi_msg msg;
+	if (display_on) {
+		msg.channel = 0x29;
+	} else {
+		msg.channel = 0x28;
+	}
+	msg.type = MIPI_DSI_DCS_LONG_WRITE;
+	msg.tx_len = 0;
+	msg.flags = 0;
+	transfer_status = mipi_dsi_transfer(priv->dsi_dev, &msg);
 	if (transfer_status != OK) {
 		lcddbg("Command %x not sent \n", cmd);
 	}
+	if (priv->mode == VIDEO) {
+		priv->config->lcd_mode_switch(true);
+	}
+	return transfer_status;
+}
+
+static int send_lcd_init_data(struct mipi_lcd_dev_s *priv)
+{
+	int transfer_status = OK;
+	struct mipi_dsi_msg msg_data, msg;
+	uint32_t size = priv->config->mipi_buffer_max_size;
+	fb_coord_t curr_row_start = 0;
+	fb_coord_t curr_row_end;
+	fb_coord_t curr_col_start = 0;
+	fb_coord_t curr_col_end = LCD_XRES - 1;
+	fb_coord_t width_size = LCD_XRES * 2;
+	uint8_t *init_buffer = (uint8_t *)kmm_malloc(priv->config->mipi_buffer_max_size);
+	if (init_buffer == NULL) {
+		lcddbg("malloc failed for %d\n", priv->config->mipi_buffer_max_size);
+		kmm_free(init_buffer);
+		return;
+	}
+	for (int i = 0; i < priv->config->mipi_buffer_max_size; i++) {
+		init_buffer[i] = 0x00;
+	}
+	switch_to_cmd(priv);
+	msg.type = MIPI_DSI_DCS_LONG_WRITE;
+	msg.tx_len = 8;
+	msg.flags = 0;
+	curr_row_end = priv->config->mipi_buffer_max_size / width_size;
+	while (curr_row_start < LCD_YRES - 1) {
+		curr_row_end = curr_row_start + (priv->config->mipi_buffer_max_size / width_size);
+		if (curr_row_end > LCD_YRES) {
+			curr_row_end = LCD_YRES;
+		}
+		if (size > (curr_row_end - curr_row_start) * LCD_XRES * 2) {
+			size = (curr_row_end - curr_row_start) * LCD_XRES * 2;
+		}
+		curr_row_end--;
+		uint8_t cmd_col_addr[8] = {0x00, (curr_col_start >> 8) & 0xFF, 0x00, curr_col_start & 0xFF, 0x00, (curr_col_end >> 8) & 0xFF, 0x00, curr_col_end & 0xFF};
+		msg.channel = 0x2A;
```

Comment:
Hello @anjana348 , I don't think the command id to configure LCD parameters are the same across different vendors. Thus, it seems not a good way to hardcode here (ie. including both channel id and cmd size).
As I understood, the command id seems vendor specific and not defined as to MIPI protocol.
Thus, could you please confirm this? If yes, we might need to change to be set dynamically in header or Kconfig.

Reply:
Hi @edwakuwaku 
I have changed logic of keeping initial garbage screen
We can keep display off till first frame comes, this doesn't need extra overhead of sending data through command mode.
Is there any issue with this way ?

Reply:
Hello @anjana348 , it should be fine. But, what I mean is in the os/drivers/lcd layer. The msg.channel is hardcoded to specific index (ie  0x28, 0x29, 0x2a, 0x2b... etc). I don't think it is good way to maintain that, because that seems vendor device specific index, instead of MIPI protocol. Which means if we switch a LCD model, this code might not be compatible.
@sunghan-chang , could you please advise here?